extractor/docker/docker_test.go

```diff
@@ -167,11 +181,175 @@ func TestExtractFiles(t *testing.T) {
 				t.Errorf("err: got %v, want %v", v.err, err)
 			}
 			if !reflect.DeepEqual(opqDirs, v.opqDirs) {
-				t.Errorf("opqDirs: got %v, want %v", opqDirs, v.opqDirs)
+				t.Errorf("OPQDirs: got %v, want %v", opqDirs, v.opqDirs)
 			}
 			if !reflect.DeepEqual(fm, v.FileMap) {
 				t.Errorf("FilesMap: got %v, want %v", fm, v.FileMap)
 			}
 		})
 	}
 }
+
+func TestDockerExtractor_SaveLocalImage(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		httpPath := r.URL.String()
+		switch {
+		case strings.Contains(httpPath, "images/get?names=fooimage"):
+			_, _ = fmt.Fprint(w, "foocontent")
+		default:
+			assert.FailNow(t, "unexpected path accessed: ", r.URL.String())
+		}
+	}))
+	defer ts.Close()
+
+	c, err := client.NewClientWithOpts(client.WithHost(ts.URL))
+	assert.NoError(t, err)
+
+	// setup cache
+	tempCacheDir, _ := ioutil.TempDir("", "TestDockerExtractor_SaveLocalImage-*")
+	defer func() {
+		_ = os.RemoveAll(tempCacheDir)
+	}()
+
+	de := DockerExtractor{
+		Option: types.DockerOption{},
+		Client: c,
+		Cache:  cache.Initialize(tempCacheDir),
+	}
+
+	r, err := de.SaveLocalImage(context.TODO(), "fooimage")
+	assert.NotNil(t, r)
+	assert.NoError(t, err)
+}
+
+func TestDockerExtractor_Extract(t *testing.T) {
+	testCases := []struct {
+		name            string
+		imageName       string
+		manifestResp    string
+		layerData       []byte
+		blobData        string
+		expectedFileMap extractor.FileMap
+		expectedError   string
+	}{
+		{
+			name: "happy path",
+			manifestResp: `{
+		  "schemaVersion": 2,
+		  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+		  "layers": [
+		     {
+		        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
+		        "size": 153263,
+		        "digest": "sha256:62d8908bee94c202b2d35224a221aaa2058318bfa9879fa541efaecba272331b"
+		     }
+		  ]
+		}`,
+			layerData: []byte{ // this is hello.txt containing "hello world\n" POSIX tar'd and then gzipped
```

Comment:
I think this tar file should include the necessary files. `hello.txt` is not a file to be extracted, so we can't test if the correct files are extracted.

Reply:
Agreed. I have made this better here: https://github.com/aquasecurity/fanal/pull/47/commits/b7565fb1533efc57e606ec3bb62fb03d6e78969a